File: leet_SLWD_0713_subproduct.cc
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k)
    {
        vector<int(*)(vector<int> &, int)> Solutions {
            Two_pointers,
        };
        return Solutions[0](nums, k);
    }

    static int Two_pointers(vector<int>& nums, int k)
    {
        if (k < 2)
            return 0;
        int res = 0;
        int prod = 1;
        int R = -1;
        int L = 0;
        while (++R < nums.size())
        {
            prod *= nums[ R ];
            while (prod >= k)
            {
                prod /= nums[ L ];
                ++L;
            }
            res += R - L + 1;
        }
        return res;
    }
};

File: leet_SLWD_0713_subproduct.go
func numSubarrayProductLessThanK(nums []int, k int) int {

    Solutions := []func([]int,int)int{
        TwoPointers,
    }
    return Solutions[0](nums, k)
}

func TwoPointers(nums[]int, k int) int {

    if k < 2 {
        return 0
    }
    res := 0
    prod := 1
    L := 0
    for R, n := range(nums) {
        prod *= n
        for prod >= k {
            prod /= nums[L]
            L++
        }
        res += R - L + 1
    }
    return res
}
